orm/encoding/ormkv/primary_key.go

```diff
@@ -0,0 +1,128 @@
+package ormkv
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/cosmos/cosmos-sdk/orm/types/ormerrors"
+
+	"google.golang.org/protobuf/proto"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// PrimaryKeyCodec is the codec for primary keys.
+type PrimaryKeyCodec struct {
+	*KeyCodec
+	msgType          protoreflect.MessageType
+	unmarshalOptions proto.UnmarshalOptions
+}
+
+// NewPrimaryKeyCodec creates a new PrimaryKeyCodec.
+func NewPrimaryKeyCodec(prefix []byte, msgType protoreflect.MessageType, fieldNames []protoreflect.Name, unmarshalOptions proto.UnmarshalOptions) (*PrimaryKeyCodec, error) {
+	keyCodec, err := NewKeyCodec(prefix, msgType.Descriptor(), fieldNames)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PrimaryKeyCodec{
+		KeyCodec:         keyCodec,
+		msgType:          msgType,
+		unmarshalOptions: unmarshalOptions,
+	}, nil
+}
+
+var _ IndexCodec = PrimaryKeyCodec{}
+
+func (p PrimaryKeyCodec) DecodeIndexKey(k, _ []byte) (indexFields, primaryKey []protoreflect.Value, err error) {
+	indexFields, err = p.Decode(bytes.NewReader(k))
+
+	// got prefix key
+	if err == io.EOF {
+		return indexFields, nil, nil
+	} else if err != nil {
+		return nil, nil, err
+	}
+
+	if len(indexFields) == len(p.fieldCodecs) {
+		// for primary keys the index fields are the primary key
+		// but only if we don't have a prefix key
+		primaryKey = indexFields
+	}
+	return indexFields, primaryKey, nil
+
+}
+
+func (p PrimaryKeyCodec) DecodeEntry(k, v []byte) (Entry, error) {
+	values, err := p.Decode(bytes.NewReader(k))
+	if err != nil {
+		return nil, err
+	}
+
+	msg := p.msgType.New().Interface()
+	err = p.Unmarshal(values, v, msg)
+
+	return &PrimaryKeyEntry{
+		Key:   values,
+		Value: msg,
+	}, err
+}
+
+func (p PrimaryKeyCodec) EncodeEntry(entry Entry) (k, v []byte, err error) {
+	pkEntry, ok := entry.(*PrimaryKeyEntry)
+	if !ok {
+		return nil, nil, ormerrors.BadDecodeEntry
+	}
+
+	if pkEntry.Value.ProtoReflect().Descriptor().FullName() != p.msgType.Descriptor().FullName() {
+		return nil, nil, ormerrors.BadDecodeEntry
```

Comment:
this error can happen for two different reasons in this method, we should wrap some more description in them

Reply:
done